interface-manager/src/interface/tap.rs

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use net::buffer::{PacketBuffer, PacketBufferMut};
+use net::interface::InterfaceName;
+use std::num::NonZero;
+use std::os::fd::AsRawFd;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tracing::{debug, error, info};
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct TapDevice {
+    file: tokio::fs::File,
+}
+
+mod helper {
+    /// This is a validated type around a value which is regrettably fragile.
+    ///
+    /// 1. Passed directly to the kernel.
+    /// 2. By a privileged thread.
+    /// 3. In an ioctl.
+    /// 4. By an implicitly null terminated pointer.
+    ///
+    /// As a result, strict checks are in place to ensure memory integrity.
+    ///
+    /// <div class=warning>
+    ///
+    /// It is essential that this type remains transparent.
+    /// Only zero-sized types may be added to this structure as we don't control the ABI.
+    /// We are subject to a contract with the kernel.
+    /// </div>
+    #[repr(transparent)]
+    #[derive(Debug, Copy, Clone)]
+    pub(super) struct InterfaceRequest(libc::ifreq);
+
+    use net::interface::InterfaceName;
+    use nix::libc;
+
+    nix::ioctl_write_ptr_bad!(
```

Comment:
I looked at the docs here for `nix` but still don't understand what a "bad" ioctl is here?  Is it one that `nix` does not understand?

Reply:
I did.  This ioctl is quite old and basically encodes the length of the structure in the ioctl identifier. This is unfortunate, but it can't be fixed :shrug:

Reply:
(Sorry I was showing the link to Manish :sweat_smile:)